lib/config/error-codes/index.js

```diff
@@ -85,8 +91,27 @@ function checkErrorCodes(errorCodes) {
         assert(Number.isInteger(errorName.code), `Error configuration file : Field 'code' must be an integer in '${domain}-${subdomain}-${error}'.`);
         assert(errorName.code > 0x0000 && errorName.code <= 0xFFFF, `Error configuration file : Field 'code' must be between 0 and 65535 in '${domain}-${subdomain}-${error}'.`);
       }
+      assert(isUnique(codesErrors), `Error configuring file : You have an error code duplicate in ${domain}-${subdomain}.`);
+      codesErrors.length = 0;
     }
+    assert(isUnique(codesSubdomains), `Error configuring file : You have a subdomain code duplicate in ${domain}.`);
+    codesSubdomains.length = 0;
   }
+  assert(isUnique(codesDomains), 'Error configuring file : You have a domain code duplicate.');
+  codesDomains.length = 0;
+}
+
+function isUnique(codesArray) {
```

Comment:
I find your implementation very complicated. 

1. You can simply use Sets instead of Arrays and in your code, before pushing code values in it, assert that it doesn't already exist
2. Use scopes: don't declare a global container just to reinitialize it each time a loop iterates, instead let JS destroy containers when exiting a scope and create a fresh new one every time you need one
3. Make the error messages show what are the duplicates: nothing is more irritating than "oh, there is an error somewhere in that file, good luck finding it" ;-)
4. You named the name of a domain `domain`, and the domain object `domainName`. This should be permutated (same remark applies to subdomains and errors).

All in all, that should give you something like that:

```js
function checkErrorCodes(errorCodesFiles) {
  const domainCodes = new Set();

  for (const domainName of Object.keys(errorCodesFiles)) {
    const domain = errorCodesFiles[domainName];

    assert(
      !domainCodes.has(domain.code), 
      `Error configuration file : domain code ${domain.code} is not unique`);

    domainCodes.add(domain.code);

    // other asserts

    const subdomainCodes = new Set();

    for (const subdomainName of Object.keys(domain.subdomains)) {
      const subdomain = domain.subdomains[subdomainName];

      assert(
        !subdomainCodes.has(subdomain.code), 
        `Error codes file : subdomain code ${subdomain.code} is not unique (domain: ${domain.code})`);

      subdomainCodes.add(subdomain.code);

      // other asserts

      const errorCodes = new Set();

      for (const errorName of Object.keys(subdomain.errors)) {
        const error = subdomain.errors[errorName];

        assert(
          !errorCodes.has(error.code), 
          `Error codes file : error code ${error.code} is not unique (domain: ${domain.code}, subdomain: ${subdomain.code})`);

        errorCodes.add(error.code);

        // other asserts
      }
    }
  }
}
```

And that way, you can remove the isUnique method.

One last remark: you should probably separate domain, subdomain and error checks in distinct functions. 
This `checkErrorCodes` is starting to be too large.

Reply:
I completely changed it right now @scottinet now

Reply:
If you separate the domain/subdomain/error checks in distinct functions, that gives something like that:

```js
function checkErrors(subdomain) {
  const errorCodes = new Set();

  for (const errorName of Object.keys(subdomain.errors)) {
    const error = subdomain.errors[errorName];

    assert(
      !errorCodes.has(error.code), 
      `Error codes file : error code ${error.code} is not unique (domain: ${domain.code}, subdomain: ${subdomain.code})`);

    errorCodes.add(error.code);

    // other asserts
  }
}

function checkSubdomains(domain) {
  const subdomainCodes = new Set();

  for (const subdomainName of Object.keys(domain.subdomains)) {
    const subdomain = domain.subdomains[subdomainName];

    assert(
      !subdomainCodes.has(subdomain.code), 
      `Error codes file : subdomain code ${subdomain.code} is not unique (domain: ${domain.code})`);

    subdomainCodes.add(subdomain.code);

    // other asserts

    checkErrors(subdomain);
  }  
}

function checkDomains(errorCodesFiles) {
  const domainCodes = new Set();

  for (const domainName of Object.keys(errorCodesFiles)) {
    const domain = errorCodesFiles[domainName];

    assert(
      !domainCodes.has(domain.code), 
      `Error configuration file : domain code ${domain.code} is not unique`);

    domainCodes.add(domain.code);

    // other asserts

    checkSubdomains(domain);
  }
}
```

And I find that far more readable.

Reply:
Why do not use this 
```

Reply:
Why do you persist to emulate a Set with an Array?

Reply:
I like you suggestion but I do not understand why you said I'm emulating Set with Array? :disappointed_relieved:

Reply:
Yoann's code pushes value to arrays, and at the end of each loop, it checks for uniqueness in each one of those arrays.
With Set, you can do the same thing, with less code, and all you have to do is to check before adding a new value to the set.

So, since both approaches seem strictly equivalent (to me at least), then I consider that using array for that purpose is like emulating sets.

Reply:
Thanks for explanations :slightly_smiling_face:

Reply:
Thanks :O